### PR TITLE
[DistillationTrainer refactor] Switch generation to GRPO's stack; delete the buffer

### PR DESCRIPTION
**Pull request by qgallouedec**
*Item 28 (Group D — generation cutover, atomic) — stacked on PR 27 (#6521). Part of #6449.*

The atomic switch from the buffer generation path to GRPO's. One PR by necessity — the buffer path and GRPO path share method names (`get_train_dataloader`, `_prepare_inputs`, samplers), so there's no green intermediate to split at. **+193 / −537** — almost all deletion; the additions are verbatim GRPO ports.

**Ported from GRPO** (with `steps_per_generation → gradient_accumulation_steps`, `num_generations → 1`, `num_iterations` dropped, VLM buffering helpers deferred to 45):
- `get_train_dataloader` / `_get_train_sampler` / `_get_eval_sampler` / `get_eval_dataloader`
- `_prepare_inputs` (buffers a generation batch, splits into `gradient_accumulation_steps` slices, reuses across the accumulation window)
- `prediction_step` (**pulled forward from item 44** — the `identity` collator makes eval batches raw lists, so GRPO's `prediction_step`, which runs `_prepare_inputs` first, is required for eval; verbatim)

**Deleted (~520 lines):** `_RepeatBatchDataLoader`, `_fill_buffer`, `_store_completions_in_buffer`, `_generate_student_completions`, `_generate_with_model`, `_build_sequence_batch`, `_get_completion_lengths`; `_DistillationCollator` → `data_collator=identity`.

**Ctor / config:** `self._step` (init here, incremented in `training_step`); dropped `num_generations` / `vllm_sync_frequency` / buffer state; `compute_loss` reads native `inputs["num_items_in_batch"]` (from `_generate_and_score_completions`, survives `split_tensor_dict`). Config: removed `num_generations` **and `generation_batch_size`** (the coupled partner of the deferred `steps_per_generation` — without it the field has one valid value, so the samplers/dataloader derive the effective batch inline). Iterable-train datasets get GRPO's ctor setup (`dispatch_batches=False` + `dataloader_num_workers=0`), required by the `identity` collator.

**Also:** brings eval generation (the eval-loss=0 limitation on #6479 is gone). Trimmed two GRPO-verbatim comments (`set_seed` rationale, the `_get_train_sampler` figure) that referenced `num_generations`/`steps_per_generation`/`num_iterations` — all removed here, so verbatim would mislead.

**Tests:** deleted `test_repeat_batch_dataloader_*` + its import; reworked the two `test_generated_batch_emits_*` to assert the GRPO keys; kept `test_num_items_in_batch_*` (premise survives).

*Review follow-ups folded in: `generation_batch_size` removed, iterable-train `dispatch_batches`/`num_workers` ctor setup, `vllm_sync_frequency` field removed.*

Verified: ruff clean; `test_distillation_trainer.py` (40) + `test_server_distillation_trainer.py` (12) green.

---

> [!NOTE]
> **Medium Risk**
> Large behavioral change to training/eval data loading and generation buffering affects on-policy distillation correctness and distributed/iterable dataset edge cases, though the diff is mostly aligned with GRPO and tests were updated.
> 
> **Overview**
> **DistillationTrainer** replaces its custom on-policy path (prompt collator, `_RepeatBatchDataLoader`, and ~500 lines of buffer/generation helpers) with the same **GRPO-style** flow: `data_collator=identity`, generation-sized train batches (`per_device × grad_accum`), `RepeatSampler` / `repeat_iterable_dataset`, and `_prepare_inputs` that calls `_generate_and_score_completions` once per accumulation window then splits with `split_tensor_dict`.
> 
> **Config/API:** Removes `num_generations`, `generation_batch_size`, `vllm_sync_frequency`, and the `data_collator` argument; loss normalization uses `inputs["num_items_in_batch"]` from generation instead of buffer state. **Iterable** train/eval datasets require `dispatch_batches=False` and train forces `dataloader_num_workers=0`. **Eval** gets a `prediction_step` override so loss runs through generation on raw prompt batches.
> 
> **Tests:** Drop `_RepeatBatchDataLoader` coverage; generation batch tests assert GRPO keys only. **ServerDistillationTrainer** drops `data_collator` from its constructor.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit dd64a6f96ad4f60ce2e2dbbd6132853f0e7c01a2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/distillation/distillation_config.py [hunk 1]
```diff
@@ -365,17 +360,13 @@ def __post_init__(self):
...
+            self.generation_batch_size = (
+                self.per_device_train_batch_size * num_processes * self.gradient_accumulation_steps
+            )
+        elif self.generation_batch_size % (self.per_device_train_batch_size * num_processes) != 0:
             raise ValueError(
-                "generation_batch_size * num_generations must equal per_device_train_batch_size * "
-                f"gradient_accumulation_steps. Got {self.generation_batch_size} * {self.num_generations} != "
-                f"{self.per_device_train_batch_size} * {self.gradient_accumulation_steps}."
+                f"generation_batch_size ({self.generation_batch_size}) must be divisible by the global batch size "
+                f"({self.per_device_train_batch_size * num_processes})."
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_config.py]: Removed generation_batch_size (05b4a13d); coupled partner of the deferred `steps_per_generation`, so it had one valid value and was a dead knob. Batch derived inline now.

File: trl/experimental/distillation/distillation_trainer.py [hunk 2]
```diff
@@ -840,256 +755,33 @@ def _generate_and_score_completions(self, inputs: list[dict[str, torch.Tensor |
...
-        inputs = self._buffered_inputs[slice_idx]
-        self._buffer_step += 1
+        mode = "train" if self.model.training else "eval"
+        if mode == "train":
+            generate_every = self.args.gradient_accumulation_steps
+            if self._step % generate_every == 0 or self._buffered_inputs is None:
+                # self._buffered_inputs=None can occur when resuming from a checkpoint
+                generation_batch = self._generate_and_score_completions(generation_batch)
+                generation_batch = shuffle_sequence_dict(generation_batch)
+                generation_batches = split_tensor_dict(generation_batch, self.args.gradient_accumulation_steps)
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:834]: answered here https://github.com/huggingface/trl/pull/6522#discussion_r3640959027

File: trl/experimental/distillation/distillation_trainer.py [hunk 3]
```diff
@@ -603,53 +467,104 @@ def _set_signature_columns_if_needed(self):
...
+            # transforming the stream instead (see `repeat_iterable_dataset`). The full permutation done by
+            # RepeatSampler becomes a buffered shuffle here.
+            if self.shuffle_dataset:
+                dataset = dataset.shuffle(seed=self.args.seed)
+            dataset = repeat_iterable_dataset(
+                dataset,
+                mini_repeat_count=1,
+                batch_size=self.args.generation_batch_size,
+                repeat_count=self.args.gradient_accumulation_steps,
+            )
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:532]: Fixed (15ae139d), ported GRPO's ctor iterable setup (dispatch_batches=False + num_workers=0). Runtime-verified.

File: trl/experimental/distillation/distillation_trainer.py [hunk 4]
```diff
@@ -603,53 +467,104 @@ def _set_signature_columns_if_needed(self):
         if self._signature_columns is None:
             self._signature_columns = ["prompt", "image", "images"]
 
-    def _get_train_sampler(self, dataset=None):
+    def get_train_dataloader(self):
+        dataset = self.train_dataset
+        if isinstance(dataset, IterableDataset):
+            # Iterable datasets can't be indexed, so RepeatSampler can't be attached. Reproduce its ordering by
+            # transforming the stream instead (see `repeat_iterable_dataset`). The full permutation done by
+            # RepeatSampler becomes a buffered shuffle here.
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:518]: see https://github.com/huggingface/trl/pull/6522#discussion_r3640966702

File: trl/experimental/distillation/distillation_config.py [hunk 5]
```diff
@@ -365,17 +360,13 @@ def __post_init__(self):
         if self.beta < 0.0 or self.beta > 1.0:
             raise ValueError(f"beta must be in [0.0, 1.0], got {self.beta}.")
 
-        if self.num_generations < 1:
-            raise ValueError(f"num_generations must be at least 1, got {self.num_generations}.")
-
-        local_sequence_batch_size = self.per_device_train_batch_size * self.gradient_accumulation_steps
+        num_processes = self.world_size
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_config.py]: Removed the field (15ae139d), its consumer (the buffer sync gate) was deleted in the cutover; syncs every step now (GRPO parity).

File: trl/experimental/distillation/distillation_trainer.py [hunk 6]
```diff
@@ -1309,32 +1036,18 @@ def _get_liger_zero3_lm_head_gather_ctx(self, model: nn.Module):
...
+        return output
 
-        return loss
+    def prediction_step(self, model, inputs, prediction_loss_only, ignore_keys: list[str] | None = None):
+        inputs = self._prepare_inputs(inputs)
+        with torch.no_grad():
+            with self.compute_loss_context_manager():
+                loss = self.compute_loss(model, inputs)
+            loss = loss.mean().detach()
+        return loss, None, None
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:1079]: Fixed in f91b4abd, `prediction_step` now wraps `compute_loss` in `_get_liger_zero3_lm_head_gather_ctx`, matching `training_step` (both removed with the Liger path at a later step).

File: trl/experimental/distillation/distillation_trainer.py [hunk 7]
```diff
@@ -1309,32 +1049,18 @@ def _get_liger_zero3_lm_head_gather_ctx(self, model: nn.Module):
...
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:1079]: Fixed in f91b4abd, restored GRPO's verbatim `self._step` / `self._buffered_inputs` comments.

**Comment by qgallouedec** [comment]
## Equivalence check: this cutover is loss-preserving

This PR replaces the buffer generation path with GRPO's samplers + `_prepare_inputs`. It changes how completions are generated and batched, and how `num_items_in_batch` is sourced, but not the loss math. Given an identical fixed batch and a matched `num_items`, `compute_loss` returns a bit-identical loss before and after the cutover, across the whole `beta` range (forward KL, JSD, reverse KL).

Result: all 21 β values identical, `max|Δ| = 0.0`.

| β | before (base, buffer) | after (PR, GRPO) |
|---|---|---|
| 0.0 (forward KL) | `0.010353946127` | `0.010353946127` |
| 0.5 (JSD) | `0.002581576118` | `0.002581576118` |
| 1.0 (reverse KL) | `0.010352961719` | `0.010352961719` |

image

Blue (before) and orange (after) lie exactly on top of each other; the bottom panel `|Δ|` is flat at 0.

### Reproduce

Run on the base branch and on this PR, then diff the two CSVs:

```bash
git checkout 27-add-generate-methods && python repro_equiv.py before.csv
git checkout 28-generation-cutover   && python repro_equiv.py after.csv
diff before.csv after.csv     # no output means identical
```

<details><summary><code>repro_equiv.py</code></summary>

```python
import sys
import torch
from trl.experimental.distillation import DistillationConfig, DistillationTrainer

OUT = sys.argv[1] if len(sys.argv) > 1 else "loss.csv"
MODEL = "trl-internal-testing/tiny-Qwen2ForCausalLM-2.5"

trainer = DistillationTrainer(
    model=MODEL, teacher_model=MODEL, args=DistillationConfig(output_dir="/tmp/eq", report_to=[])
)

# Perturb the teacher (seeded, identical on every branch) so student != teacher and the JSD loss is non-trivial.
g = torch.Generator().manual_seed(0)
with torch.no_grad():
    for p in trainer.teacher_model.parameters():
        p.add_(0.05 * torch.randn(p.shape, generator=g).to(p.device, p.dtype))

# One fixed batch, reused for every beta.
vocab = trainer.model.config.vocab_size
dev = trainer.accelerator.device
tg = torch.Generator().manual_seed(1)
B, P, C = 3, 5, 7
prompt_ids = torch.randint(0, vocab, (B, P), generator=tg).to(dev)
completion_ids = torch.randint(0, vocab, (B, C), generator=tg).to(dev)
prompt_mask = torch.ones(B, P, dtype=torch.long, device=dev)
completion_mask = torch.ones(B, C, dtype=torch.long, device=dev)
completion_mask[0, -2:] = 0
completion_mask[2, -1] = 0
num_items = completion_mask.sum()

trainer._buffered_num_items = num_items  # num_items source on the base (buffer); ignored on the PR
trainer.model.train()                    # compute_loss only applies the num_items normalizer in training mode

with open(OUT, "w") as f:
    f.write("beta,loss\n")
    for i in range(21):  # beta = 0.0 (forward KL) .. 0.5 (JSD) .. 1.0 (reverse KL)
        trainer.beta = i / 20
        batch = {
            "prompt_ids": prompt_ids,
            "prompt_mask": prompt_mask,
            "completion_ids": completion_ids,
            "completion_mask": completion_mask,
            "num_items_in_batch": num_items,  # num_items source on the PR (GRPO); ignored on the base
        }
        with torch.no_grad():
            loss = trainer.compute_loss(trainer.model, batch)
        f.write(f"{trainer.beta:.3f},{loss.item():.12f}\n")
print("wrote", OUT)
```
</details>

File: trl/experimental/distillation/distillation_trainer.py [hunk 8]
```diff
@@ -603,53 +504,117 @@ def _set_signature_columns_if_needed(self):
...
+        # The generation batch is the full effective training batch. GRPO names it `generation_batch_size` (paired with
+        # the deferred `steps_per_generation`); both are deferred here, so derive it inline — slot-in point.
+        generation_batch_size = (
+            self.args.per_device_train_batch_size
+            * self.accelerator.num_processes
+            * self.args.gradient_accumulation_steps
+        )
         return RepeatSampler(
             data_source=dataset,
-            mini_repeat_count=self.num_generations,
```
Comment (qgallouedec) [inline trl/experimental/distillation/distillation_trainer.py:611]: self.num_generations is always one, and the arg is removed in this PR

File: trl/experimental/distillation/distillation_trainer.py [hunk 9]
```diff
@@ -603,53 +504,117 @@ def _set_signature_columns_if_needed(self):
...
-            batch_size=self.args.generation_batch_size * self.accelerator.num_processes,
```
Comment (qgallouedec) [inline trl/experimental/distillation/distillation_trainer.py:612]:
```
batch_size = self.args.generation_batch_size * self.accelerator.num_processes
           = self.per_device_train_batch_size * self.gradient_accumulation_steps // num_generations * self.accelerator.num_processes
```
see https://github.com/huggingface/trl/pull/6522/changes#diff-d434e0fdf58b21840143966d2cce6cd4a1f151c681b534eb845f031a6aa1d775L373

then, since `num_generations==1`

```
batch_size = self.args.generation_batch_size * self.accelerator.num_processes
           = self.per_device_train_batch_size * self.gradient_accumulation_steps * self.accelerator.num_processes
```

File: trl/experimental/distillation/distillation_trainer.py [hunk 10]
```diff
@@ -603,53 +504,117 @@ def _set_signature_columns_if_needed(self):
...
-            repeat_count=1,
-            shuffle=True,
+            mini_repeat_count=1,
+            batch_size=generation_batch_size,
+            repeat_count=self.args.gradient_accumulation_steps,
```
Comment (qgallouedec) [inline trl/experimental/distillation/distillation_trainer.py:558]: `repeat_count=self.args.gradient_accumulation_steps` implements the "generate once per optimizer step, reuse across the accumulation window."

see here https://github.com/huggingface/trl/pull/6522/changes#diff-03c25c1f305d4a0687da7ce7207f80fe582927a92dd37ae8efd5b80341b4bc1cR823

File: trl/experimental/distillation/distillation_trainer.py [hunk 11]
```diff
@@ -840,256 +813,35 @@ def _generate_and_score_completions(self, inputs: list[dict[str, torch.Tensor |
 
     @profiling_decorator
     def _prepare_inputs(self, generation_batch: dict[str, torch.Tensor | Any]) -> dict[str, torch.Tensor | Any]:
-        if not self.model.training:
-            return generation_batch
-
-        buffer_steps = self.args.gradient_accumulation_steps
-        if self._buffer_step % buffer_steps == 0 or self._buffered_inputs is None:
-            self._fill_buffer(generation_batch, buffer_steps)
+        # Prepares inputs for model training/evaluation by managing completion generation and batch handling.
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:814]: This matches GRPO: its `_generate_and_score_completions` also writes completions into the shared buffer unconditionally, and eval generation here is the intended GRPO-parity behavior. The train/eval separation comes from GRPO's `log()` + bounded `_logs` deques, which distillation adopts in the `_textual_logs` → `_logs` conversion (item 43 of #6449). Keeping the write GRPO-consistent until then rather than adding a distillation-only guard.

File: trl/experimental/distillation/distillation_trainer.py [hunk 12]
```diff
@@ -603,53 +504,125 @@ def _set_signature_columns_if_needed(self):
...
+            dataset = repeat_iterable_dataset(
+                dataset,
+                mini_repeat_count=1,
+                batch_size=generation_batch_size,
+                repeat_count=self.args.gradient_accumulation_steps,
+            )
+        return self._get_dataloader(
+            dataset=dataset,
+            description="Training",
+            batch_size=self._train_batch_size * self.args.gradient_accumulation_steps,  # < this is the change
```
Comment (cmpatino) [inline trl/experimental/distillation/distillation_trainer.py:536]: We use `self._train_batch_size` here and `per_device_train_batch_size` in the sampler code. 

The distillation trainer allows `auto_find_batch_size`, so `self._train_batch_size` might change while `per_device_train_batch_size` remains static. If the two values are different, we may have the wrong dynamics in the interplay between the dataloader and the sampler.

What we could do is replicate what we do in the GRPO trainer and simply disable `auto_find_batch_size` altogether https://github.com/huggingface/trl/blob/190c39c73491c41a4e24e91799da6440097bb00f/trl/trainer/grpo_config.py#L1065)

File: trl/experimental/distillation/distillation_trainer.py [hunk 13]
```diff
@@ -418,13 +323,43 @@ def __init__(
...
+            args.dataloader_num_workers = 0
+
         # Trainer does not need to remove unused columns — the collator handles raw data
         args.remove_unused_columns = False
 
         super().__init__(
             model=model,
             args=args,
-            data_collator=data_collator,
+            data_collator=identity,  # No data collation is needed in distillation
```
Comment (cmpatino) [inline trl/experimental/distillation/distillation_trainer.py:360]: I agree with the change, but we should also remove the data_collator param from the ctor. That way we establish upfront that we don't use a data collator at all.

Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:360]: that's a good point!

Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:360]: https://github.com/huggingface/trl/pull/6522/commits/d497c6061d50717bf5c7425caafb50a68267b651

File: trl/experimental/distillation/distillation_trainer.py (hunk 12 above)
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:536]: indeed, I'll make sure we replicate this recent guard in a future pr, thanks for raising

File: trl/experimental/distillation/distillation_trainer.py [hunk 14]
```diff
@@ -603,53 +502,125 @@ def _set_signature_columns_if_needed(self):
...
+            * self.args.gradient_accumulation_steps
+        )
         return RepeatSampler(
             data_source=dataset,
-            mini_repeat_count=self.num_generations,
-            batch_size=self.args.generation_batch_size * self.accelerator.num_processes,
-            repeat_count=1,
-            shuffle=True,
+            mini_repeat_count=1,
+            batch_size=generation_batch_size,
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:557]: yes, done in a future pr, same as https://github.com/huggingface/trl/pull/6522#discussion_r3657161021

File: trl/experimental/distillation/distillation_config.py [hunk 15]
```diff
@@ -330,10 +312,6 @@ class DistillationConfig(_BaseConfig):
         default=None,
         metadata={"help": "Regex pattern for vLLM structured outputs."},
     )
-    vllm_sync_frequency: int = field(
-        default=1,
-        metadata={"help": "Frequency (in training steps) to synchronize student weights to the vLLM engine."},
-    )
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_config.py:336]: I know! third time aha
